### ISSUE TITLE
Repair the treadmill controller

### ISSUE THREAD

**Issue by moorepants** [mechmotum/issues#1]
The treadmill controller is kaput. We need to try to repair it or make/find a replacement. We could start with the meetshop and see if they can fix it or maybe we can pay DEMO to fix it. Asking Jan van Frankenhuizen for suggestions on getting it fixed, maybe he can help or point us to a person in 3mE that can do this.

**Comment by moorepants** [comment]
We need the treadmill working for January 2024 for Marten's experiments. Here are some notes on what we can do.

- There is a working controller on the second treadmill in the basement. This, as well as the connection wire on our treadmill should be able to be used to determine what the pins do.
- We can try to arrange to share the working controller with the lab in the basement. Mostly just needs coordination as it is unlikely experiments will all be at the same time in both labs. Maybe we can even borrow if for a long period.
- A student group tapped into our treadmill with an ardiuno to automatically control the speed. maybe we can utilize that setup to manually control the speed. We need an emergency stop button working and speed control. It should be possible to make our own rudimentary controller.

**Comment by moorepants** [comment]
There is a sticker on the other lab's controller that says "Motek Medical service date 2020". Given that they were servicing it up to 2020, we should call Motek. Motek bought forcelink at some point. @pcvanderzalmtudelft

**Comment by pcvanderzalmtudelft** [comment]
That is a good idea, I will try to contact them this afternoon.

**Comment by pcvanderzalmtudelft** [comment]
They didnt answer my phone call, so I have send them an email, with @moorepants in cc (although the email is in dutch). Lets hope they respond soon, but I will try to call again tomorrow.

**Comment by pcvanderzalmtudelft** [comment]
I have called Motek again and they said that it would be possible to buy a new controller. I have send them another email with photos of the controller we have now with the serial number to see if they have a matching one laying around.

**Comment by moorepants** [comment]
Perfect, if they have it let me know the price.

**Comment by pcvanderzalmtudelft** [comment]
Motek said that these controllers are pretty old and they hope that they can still support them; I have had to send photos of the PCB and they will now internally discuss the options.

**Comment by moorepants** [comment]
Ok, any replacement of something new or used is fine too.

**Comment by moorepants** [comment]
The new board is here but two of the ribbon cables need new connectors or new cables for the new board. Also seems one chip may need to be transferred from the old board to the new board. Marten will take it to the meetshop in the morning for advice.

**Comment by moorepants** [comment]
From Marten:

> I just brought the board to the meetshop. He said the connector on the new board is also for a ribbon cable. He is going to check if he has the right connector, and will otherwise order it. He will let me know when he has an update.

**Comment by pcvanderzalmtudelft** [comment]
I got a message from Jacques that the controller is fixed and I will pick it up today

**Comment by moorepants** [comment]
Woohoo!

**Comment by pcvanderzalmtudelft** [comment]
Later today (or tomorrow, depending on the availability of Jacques) I will go over to the meetshop to pick up the controller and ask him for suggestions for the next steps. I will also taka look at the forum, although it seems that at the sales section, there is not much traffic.

**Comment by moorepants** [comment]
Doesn't' hurt to make some traffic. Lots of people still watch that forum.

**Comment by moorepants** [comment]
Summary of findings from June 6, 2025:

- The original board in the user controller is BONTE B.V. ZWOLLE 11-0082, TREADMILL CONTR. V1.01 with the Enigma B.V. Zwolle 11-0082 V 0.6.0 chip in the U11 slot. The back of the metal enclosing has a sticker: EN-BO Systems MODEL DISPLAY PANEL PROGRAMMABLE HV SERIAL NR. TMB-01-011.
- Motek medical sent us a replacement board labeled BONTE B.V. ZWOLLE 11-0082, TREADMILL CONTR. V1.01 with no chip in the U11 slot.
- We got the controller back from Jacques with the new board installed and the Enigma chip from the old board.
-  We also have a working controller from the basement. This is also a BONTE B.V. ZWOLLE 11-0082, TREADMILL CONTR. V1.01 but it had a Enigma V0.65 chip.
- The controller from the basement works fine with our treadmill.
- We tried the V0.65 chip in the replacement board in our controller, but the 888888 lights just come on and nothing responds to the buttons.
- We also tried the V0.60 chip in the replacement board (the way Jacques returned it) that behaved the same as the V0.65 chip in the replacement board. Just lighted 888888.
- We then installed the original board with the V.60 chip but did not quite connnect the large strip cable correctly.We plugged it in and Bart saw some LEDs flash, but no LEDs numbers lit.  The system squealed from the tiny speaker for a while.  We then corrected the strip cable and plugged it in. No lights at all this time. It seemed completely dead at this point. Bart also saw a light flash in the treadmill when we connected the old board the first time (and it squealed). So maybe we just did more damage to the original board.
- The old board with the v0.6.0 chip is installed in the controller.

**Comment by chris-konrad** [comment]
# Building our own controller

In the last session, we found the old BEP project that controlled the treadmill with an Arduino. We can use the same approach to build our own controller.

## Electrical Design (Christoph):
- Arduino
- 8-segment or LCD display for current speed
- Start/Stop button
- Up/Down pushbuttons for speed control 
- Emergency Stop function
- Status LEDs or: Running - Stand-by - Emergency-Stop
- Serial interface for communication with the treadmill 
- Power supply via treadmill 

## Software Design (Neville)
Repurpose BEP code:
- Read-out start button and send start command via serial
- Read-out speed buttons and send speed commands via serial, display current speed
- Read-out emergency status and display emergency state. 
- Control Running and Stand-by status LEDs

## Hardware (Bart)
Build controller box.

## Next steps:
1. (Christoph) Determine interface: What communication protocol (I2C, SPI) does the treadmill use? What power does the treadmill supply. How is the emergency function connected? 
2. (Christoph) Design electrics and make part-list. 
4. (Neville) Build software
5. (Bart) Design housing and make part-list.
6. (Rawdha) Acquire parts
7. (All) Set up on breadboard and test.
8. (All) Final assembly.

**Comment by Neville-N** [comment]
@chris-konrad As far as I understand, the communication is done with UART ( the RX and TX pins on the Arduino) with the baud rate set to 19200.

**Comment by moorepants** [comment]
@Neville-N it would be good to make a new repo in the mechmotum org and place the bep code in it. we can then move design conversations and such to the issue tracker of that repo. can you do that?

**Comment by chris-konrad** [comment]
Yes, lets do that! But lets just call it "treadmill-controller" or something, not the name the BEP project. It will quickly evolve from the BEP material to its new purpose. Would be nice to have everything related to the development in one repo.

**Comment by Neville-N** [comment]
The repo is in the air at https://github.com/mechmotum/treadmill-controller
I made it private to be sure, but do all lab members still have access?

**Comment by moorepants** [comment]
We should make it public so everyone can see it.

**Comment by Neville-N** [comment]
It's public now

**Comment by moorepants** [comment]
The controller is fixed and mounted on the control pole. We only need to make a back cover for the 3D printed box so that the wires are not exposed.

**Comment by moorepants** [comment]
Back plate noted here: https://github.com/mechmotum/treadmill-controller/issues/11, closing this issue as complete. Details on the treadmill controller repo can still be addressed, but main issue is done! Only took us 3 years!
